### PR TITLE
feat(GiniHealthSDK): Payment providers added to payment component IPC-97

### PR DESCRIPTION
**Pull request by razvancapra**
Description: https://ginis.atlassian.net/browse/IPC-97

### CODE REVIEW

File: HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentComponentView.swift [hunk 1]
```diff
@@ -68,31 +73,32 @@ final class PaymentComponentView: UIView {
...
     }()
     
     private lazy var selectBankPickerView: UIView = {
         let view = UIView()
         view.translatesAutoresizingMaskIntoConstraints = false
-        view.frame = CGRect(x: 0, y: 0, width: .max, height: 56)
-        view.layer.cornerRadius = 12
+        view.frame = CGRect(x: 0, y: 0, width: .greatestFiniteMagnitude, height: Constants.selectBankViewHeight)
+        view.layer.cornerRadius = Constants.cornerRadius
         view.layer.borderWidth = 1
```
Comment (zladzeyka) [inline HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentComponentView.swift:102]: [Suggestion]: @razvancapra could you please also move `borderWidth` in the Constants?

File: HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentComponentView.swift [hunk 2]
```diff
@@ -165,105 +172,92 @@ final class PaymentComponentView: UIView {
...
+            selectBankPickerView.addSubview(bankImageView)
+        }
+        selectBankPickerView.addSubview(bankNameLabel)
+        selectBankPickerView.addSubview(chevronDownIconView)
+        payInvoiceView.addSubview(payInvoiceLabel)
+
+        activateConstraints()
+    }
+
+    private func activateConstraints() {
```
Comment (zladzeyka) [inline HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentComponentView.swift:200]: [Clarification]: @razvancapra can we refactor it even in more granular methods separating by view for example, so far for this method we have 46 lines 😅?

File: HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentComponentViewModel.swift [hunk 3]
```diff
@@ -29,54 +30,94 @@ final class PaymentComponentViewModel {
...
-                                                                              size: 14,
-                                                                              style: .linkBold)
-        self.selectBankLabelFont = GiniHealthConfiguration.shared.customFont.with(weight: .medium,
-                                                                     size: 14,
-                                                                     style: .subtitle2)
-        self.bankImageIconName = bankIconName
-        self.bankNameLabelText = bankName
+                                                                                           size: 14,
+                                                                                           style: .linkBold)
+        self.selectYourBankLabelFont = GiniHealthConfiguration.shared.customFont.with(weight: .medium,
```
Comment (zladzeyka) [inline HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentComponentViewModel.swift:107]: [Clarification]: why we don't use here GiniHealthConfiguration.shared.textStyleFonts[.subtitle2] instead?

File: HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentComponentViewModel.swift [hunk 4]
```diff
@@ -29,54 +30,94 @@ final class PaymentComponentViewModel {
...
-         bankIconName: String, 
-         payInvoiceAccentColor: GiniColor,
-         payInvoiceTextColor: GiniColor,
+    init(paymentProvider: PaymentProvider?,
          giniHealth: GiniHealth) {
         self.giniHealth = giniHealth
         self.moreInformationLabelFont = GiniHealthConfiguration.shared.customFont.with(weight: .regular,
-                                                                          size: 13,
-                                                                          style: .caption1)
+                                                                                       size: 13,
```
Comment (zladzeyka) [inline HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentComponentViewModel.swift:102]: [Clarification]: why we don't use here GiniHealthConfiguration.shared.textStyleFonts[.caption1] instead?

File: HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentComponentsController.swift [hunk 5]
```diff
@@ -34,19 +55,34 @@ public final class PaymentComponentsController: NSObject, PaymentComponentContro
 
     var paymentComponentView: PaymentComponentView!
 
+    public func getPaymentProviders(completion: @escaping (Result<PaymentProviders, GiniHealthError>) -> Void) {
```
Comment (zladzeyka) [inline HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentComponentsController.swift:58]: [Suggestion]: could you please move the implementation in `GiniHealth`?

**Review by zladzeyka — CHANGES_REQUESTED** [review]
Thank you @razvancapra, great job!👏🏻
There are some small comments and I would like to ask to sort out the extensions. Let's keep the extensions folder and move the existing there as well
Screenshot 2024-02-06 at 11 56 28
I think you `UIColor`extension can be moved in UIColor + Utils and etc.

**Review by zladzeyka — APPROVED** [review]
Thank you!

**Comment by ValentinaIancu-Gini** [comment]
Thank you @razvancapra for your great work!

I took the liberty to test on my favourite device 🙈 : iPhone SE first generation with iOS 15.5(simulator) and I saw that the name of the provider did not fit in the label. Please see the image attached.
Screenshot 2024-02-09 at 11 19 31

Also in dark mode, the dropdown arrow is not very visible:

Simulator Screenshot - iPhone 15 Pro Max - 2024-02-09 at 11 25 20

**Comment by ValentinaIancu-Gini** [comment]
@razvancapra thanks for fixing those UI issues. Now is looking good both on dark mode and on that small device! Great job!

File: HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentComponentsController.swift [hunk 6]
```diff
@@ -8,22 +8,44 @@
...
+        DispatchQueue.main.async {
+            self.isLoading = true
+            self.giniHealth.checkIfAnyPaymentProviderAvailable { [weak self] result in
+                self?.isLoading = false
+                switch result {
+                case let .success(paymentProviders):
+                    self?.paymentProviders = paymentProviders
+                    self?.delegate?.didFetchedPaymentProviders(paymentProviders)
+                case let .failure(error):
+                    self?.delegate?.didReceivedErrorOnPaymentProviders(error)
```
[the remark this answers is not in the document]
Comment (razvancapra) [reply HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentComponentsController.swift]: I removed handling this error. We only print a message in the debugger.

File: HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentComponentsController.swift [hunk 7]
```diff
@@ -8,22 +8,44 @@
...
+    public weak var viewDelegate: PaymentComponentViewProtocol?
 
     private var giniHealth: GiniHealth
-    var paymentProviders: PaymentProviders = []
+    private var paymentProviders: PaymentProviders = []
 
     public init(giniHealth: GiniHealth) {
         self.giniHealth = giniHealth
+        super.init()
+        DispatchQueue.main.async {
```
[the remark this answers is not in the document]
Comment (razvancapra) [reply HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentComponentsController.swift]: After productive calls, we've decided to add a secondary function called: `loadPaymentProviders` that needs to be called by the SDK consumer. Thanks for suggestion

File: HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentPrimaryButton.swift [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+//  Copyright © 2024 Gini GmbH. All rights reserved.
+//
+
+
+import UIKit
+import GiniHealthAPILibrary
+
+final class PaymentPrimaryButton: UIView {
+    
+    private var giniHealthConfiguration = GiniHealthConfiguration.shared
```
Comment (zladzeyka) [inline HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentPrimaryButton.swift:13]: [Clatification]:  do we use `giniHealthConfiguration` here?

**Review by zladzeyka — COMMENTED** [review]
Thank you @razvancapra 🙏,
 left a few comments

File: HealthSDK/GiniHealthSDKExample/GiniHealthSDKExample/InvoicesList/InvoicesListViewModel.swift [hunk 9]
```diff
@@ -102,11 +72,34 @@ final class InvoicesListViewModel {
         self.invoices = invoices ?? hardcodedInvoicesController.getInvoicesWithExtractions()
         self.documentService = documentService
         self.paymentComponentsController = paymentComponentsController
+        self.paymentComponentsController.delegate = self
     }
     
     func viewDidLoad() {
+        paymentComponentsController.loadPaymentProviders()
```
[the remark this answers is not in the document]
Comment (razvancapra) [reply HealthSDK/GiniHealthSDKExample/GiniHealthSDKExample/InvoicesList/InvoicesListViewModel.swift:79]: we call it in the InvoicesListViewController, this is just a helping function from viewModel called at the same time. T

File: HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentPrimaryButton.swift (hunk 8 above)
Comment (razvancapra) [reply HealthSDK/GiniHealthSDK/Sources/GiniHealthSDK/Core/PaymentComponent/PaymentPrimaryButton.swift:13]: yes, for the button font: `giniHealthConfiguration.textStyleFonts[.button]`

**Review by zladzeyka — APPROVED** [review]
Looks good to me 😇
We can prepare a build for QA 🙏

**Pull request merged**